### PR TITLE
[WIP] task stealing mechanism for thread pool better performance

### PR DESCRIPTION
**Pull request by RishabhRD**
## Current implementation of threadpool looks as follows

There is a huge problem with this implementation: Contention.

Every-time a thread would try to push or pop a task, it needs to acquire a mutex. If another thread is pushing or popping at same time, we need to block. This blocking leads to context switching kills performance and many other tasks could be done at this time. Too much context switching can degrade performance as it is like to have a single thread.

## Enhancement 1

Instead of blocking for a single task_queue, each thread would have their own task_queue. With this, there is no need to block if other thread is popping the result and we are pushing to some other queue. We can push task in simple round robin fashion. This looks as follows:

## Problem with enhancement 1
It can be possible that, some long running task are getting accumulated to one task queue and other task queue are just sitting idle. Also, still it is possible that a thread is trying to pop from its task queue and we are pushing to same task queue and that point that thread needs to block and this kills performance.

## Proposed Solution (Enhancement 2)
We can implement task stealing so that no task queue bubble up tasks with remaining task queue sitting idle. The main idea here is when thread needs to pop a task from task_queue, it would "try" to pop. If it is not able to pop(because someone else is popping and mutex is acquired or queue is empty) it moves to next task queue. If in a cycle it was not able to pop, then only it blocks for its queue.
Also while pushing to queue, it "try" to push to every task queue. If it is not able to push because some other thread is holding that mutex, then it tries to push to other task queue. If in a cycle it was not able to push to any queue, then only it blocks for its queue.

This looks as follows:

This solution ensures that, a thread is spending least time in blocking. And thus in context switching.

## Implementation challenges
This solution currently doesn't compile. The reasons goes as follows. My previous PR was regarding threadpool should not be a singleton. I have a notification_queue class that is abstraction for task queue. Now, we need ``std::conditional_variable`` as a data member of notification_queue for blocking till queue is empty or we are done.

``std::conditional_variable`` is not copy and move constructible (obviously it should not be). This leads to notification_queue not to be copy and move constructible. It should not be too as its acting as a resource for a thread processing.

Talking about ``Initialize`` method of threadpool, it needs to initialize ``std::vector<notification_queue> m_q``.
There are 2 ways of doing this:
- ``m_q.resize(count_threads)``
- ``m_q.clear()`` and then for count_thread times push/emplace_back notification_queue to m_q

With any of 2 approach, notification_queue needs to be copy/move constructible: resize for obvious reasons, and in second approach push_back and emplace_back needs that.

This happens because, initialization of data structures happen at time of ``static ThreadPool thread_pool``. When we call our ``Initialize`` function, it modifies the data structure not actually initialize it. (Assignment and construction are 2 very different operation inspite of having very similar syntax). This modification needs copy/move construction but initialization of m_q in constructor would have done that in-place that would not need any copy/move of notification_queue.

I would comment on part of code that isn't compilable for the time and then we can discuss what measures we can take.

### CODE REVIEW

File: p4-fusion/thread_pool.cc [hunk 1]
```diff
@@ -83,58 +71,53 @@ void ThreadPool::Resize(int size)
 	Initialize(size);
 }
 
-void ThreadPool::Initialize(int size)
+void ThreadPool::Initialize(unsigned size)
 {
+	m_count = size;
+	m_index = 0;
+	m_q.resize(m_count);
```
Comment (RishabhRD) [inline p4-fusion/thread_pool.cc:78]: This needs notification_queue to be copyable or movable and that's not possible as std::conditional_variable is one data member of that class and thus this doesn't compile. (Discussed in topic)

**Comment by ChronicallySerious** [comment]
Thanks for sending the PR. I do feel that this is a worthy change to have and make efforts to perfect. In the end, we can measure exactly how much performance we are improving and if it is a considerable change then likely merge it into master down the line after any lingering bugs or crashes are fixed.

The approach you mention is worth a try. 

---

To work around the compilation issue, you might take hints from the `ChangeList` class.

https://github.com/salesforce/p4-fusion/blob/master/p4-fusion/commands/change_list.h

This class was required to contain non-copyable or move-constructible elements and it also needs to be stored in a dynamic vector, so you might observe that all the data members that are not move-constructible/copyable are stored as objects beneath a std::shared_ptr and not as direct members of this class. By doing this, we make objects of the `ChangeList` class moveable since std::shared_ptr has fully defined move semantics.

However, we do make sure to never *actually* move a `ChangeList` object in our code, this is just a workaround to be able to store them in a dynamic array, which is not really resized once it is initialized. That is by intention.

File: p4-fusion/thread_pool.cc (hunk 1 above)
Comment (ChronicallySerious) [reply p4-fusion/thread_pool.cc:78]: You can refer my latest comment under the thread to fix this

**Comment by ChronicallySerious** [comment]
Also, I should mention that we use PascalCase for class names, camelCase for local variables, and m_PascalCase for class member variables. Please adhere to these project conventions.

**Comment by RishabhRD** [comment]
I made ``std::mutex`` and ``std::conditional_variable`` under a ``std::unique_ptr``, hence now class is movable and thus it compiles. I have fixed naming conventions whereever I noticed.

Btw I am still concerned of dynamic allocation of mutex and conditional_variable as those are really unnecessary and now we have to make sure that we don't do move, that should have been the job of compiler. But that needs a different kind of refactor and we can look at those issues in any different PR. I guess we are more focusing on performance aspect of it in this PR.

**Comment by ChronicallySerious** [comment]
@RishabhRD Do you have metrics on how much performance this is gaining?

**Comment by RishabhRD** [comment]
@ChronicallySerious not exactly for this project. I have profiled the code for my other personal project and it gained a high performance. But that was a different usecase with much CPU-bound tasks.
I am not seeing any benchmark directory or something related to it here or even dependency to any benchmarking library. What steps are taken to profile the code? I would go through the same process to present the benchmark report.

**Comment by ChronicallySerious** [comment]
@RishabhRD It would be extremely helpful to test this with a codebase like ours where the most timing taking Threadpool tasks are mostly networking I/O bound and not CPU bound.

To perform a benchmark, we can make use of the profiling feature that is integrated into p4-fusion. Take a look at the note in the README about the `--trace` build option. After that, you would have to -

1. Build p4-fusion with the `--trace` option but excluding the changes in this PR
2. Pick a Perforce depot path to test this on
3. Run p4-fusion on that path and take note of the final logs the p4-fusion prints (it contains information about the time it took to process the depot and the amount of processing was done)
4. Copy the `trace.json` file that is generated at the repository path to someplace safe.
5. Build p4-fusion with the `--trace` option and also include the changes in this PR
6. Repeat the same p4-fusion command and take note of the final logs and the `trace.json` file
7. Send the benchmarks results here along with the two JSON files. Please make sure to exclude any sensitive information or any company-related information out of the JSON files or the logs.

**Comment by RishabhRD** [comment]
Hi @ChronicallySerious I just started to do the benchmarking thing, however I am very new to this P4 depot thing. I exactly have no knowledge of it :sweat_smile: . To run the application I need these parameters:
- path : Seems like its a path to some p4 project, but I don't have any such project and I searched on internet but don't got any
- src : Seems like this is where the cloning would be done, so no problem with that.
- client: No idea what does this mean, seems like something related to p4
- lookAhead : Some important value, seems like I should use 2000 form README
- port : Again the same no idea
- user : looks similar to git user

Any hints where I can look for them. I have just joined salesforce few days back, so we can contact on slack for this too, if there is some data we can't share here publically.

**Comment by RishabhRD** [comment]
I removed the Wait method from ThreadPool class for the time being. I would reimplement it to support the performance of new thread pool model. It should not be a problem in building project as it is unused.

## Problem with Wait method

I realized that we are using ``std::atomic<long> m_JobsProcessing`` for tracking number of jobs being currently processed (or not finished). And wait method would busy wait for all jobs processing (to achieve kind of when_all async algorithm behavior). Everytime a thread completes a task, it accesses the same shared atomic and also while adding job this is the same case. Contending thread on atomic is not good for performance. (There are many platform dependent reasons for this). This model would give us the similar behavior where we started.

## Possible solution
We can again go with ``std::vector<jobs_processing>`` in a very similar way how we made ``notification_queue`` to be unique for each thread. There are other design choices too like, notification_queue itself can maintain the count. I would come back on this, after some local experiments.

**Comment by twarit-waikar** [comment]
It is usually best to remove functions that are no longer used and spend time on other things that do get utilized. 

Also, you have already explained how the Wait function *would* work in your new threadpool so later on, if we feel like we need it, then there is no issue in referring back to your comment and using that as a reference for the actual implementation.

**Comment by twarit-waikar** [comment]
Closing since this PR has gotten stale. Please reopen this in case this is still being worked upon.

**Pull request closed without merging**